File: php/src/SlhDsa/Hypertree.php
<?php

declare(strict_types=1);

namespace PQC\SlhDsa;

/**
 * Hypertree for SLH-DSA.
 * FIPS 205 Section 7.
 */
final class Hypertree
{
    /**
     * Generate hypertree signature.
     * FIPS 205 Algorithm 11.
     *
     * @param string $msg Message (n bytes - typically FORS pk)
     * @param string $skSeed Secret seed
     * @param string $pkSeed Public seed
     * @param int $idxTree Tree index
     * @param int $idxLeaf Leaf index in bottom tree
     * @param array $params Parameters
     * @return string Hypertree signature
     */
    public static function sign(
        string $msg,
        string $skSeed,
        string $pkSeed,
        int $idxTree,
        int $idxLeaf,
        array $params
    ): string {
        $d = $params['layers'];
        $hPrime = $params['tree_height'];
        $n = $params['n'];

        $adrs = new Address();

        // Sign at the bottom layer
        $adrs->setLayerAddress(0);
        $adrs->setTreeAddress($idxTree);

        $sig = Xmss::sign($msg, $skSeed, $pkSeed, $idxLeaf, $adrs, $params);
        $root = Xmss::rootFromSig($msg, $sig, $pkSeed, $idxLeaf, $adrs, $params);

        // Sign at each subsequent layer
        for ($j = 1; $j < $d; $j++) {
            $idxLeaf = $idxTree & ((1 << $hPrime) - 1);
            $idxTree >>= $hPrime;

            $adrs->setLayerAddress($j);
            $adrs->setTreeAddress($idxTree);

            $sig .= Xmss::sign($root, $skSeed, $pkSeed, $idxLeaf, $adrs, $params);
            if ($j < $d - 1) {
                $root = Xmss::rootFromSig($root, substr($sig, -($hPrime + $params['len']) * $n), $pkSeed, $idxLeaf, $adrs, $params);
            }
        }

        return $sig;
    }

    /**
     * Verify hypertree signature.
     * FIPS 205 Algorithm 12.
     */
    public static function verify(
        string $msg,
        string $htSig,
        string $pkSeed,
        int $idxTree,
        int $idxLeaf,
        string $pkRoot,
        array $params
    ): bool {
        $d = $params['layers'];
        $hPrime = $params['tree_height'];
        $n = $params['n'];
        $len = $params['len'];

        $adrs = new Address();
        $xmssSigSize = ($hPrime + $len) * $n;

        // Verify at the bottom layer
        $adrs->setLayerAddress(0);
        $adrs->setTreeAddress($idxTree);

        $xmssSig = substr($htSig, 0, $xmssSigSize);
        $node = Xmss::rootFromSig($msg, $xmssSig, $pkSeed, $idxLeaf, $adrs, $params);

        // Verify each subsequent layer
        for ($j = 1; $j < $d; $j++) {
            $idxLeaf = $idxTree & ((1 << $hPrime) - 1);
            $idxTree >>= $hPrime;

            $adrs->setLayerAddress($j);
            $adrs->setTreeAddress($idxTree);

            $xmssSig = substr($htSig, $j * $xmssSigSize, $xmssSigSize);
            $node = Xmss::rootFromSig($node, $xmssSig, $pkSeed, $idxLeaf, $adrs, $params);
        }

        return hash_equals($node, $pkRoot);
    }
}

File: java/src/main/java/com/pqc/slhdsa/Hypertree.java
package com.pqc.slhdsa;

/**
 * Hypertree operations per FIPS 205 Section 7.
 */
public final class Hypertree {

    private Hypertree() {}

    /**
     * ht_sign(M, SK.seed, PK.seed, idx_tree, idx_leaf):
     * Generate a hypertree signature (Algorithm 15 in FIPS 205).
     */
    public static byte[] sign(byte[] m, byte[] skSeed, byte[] pkSeed,
                               long idxTree, int idxLeaf,
                               SlhParams params, SlhHash hash) {
        int n = params.n;

        Address adrs = new Address();
        adrs.setLayerAddress(0);
        adrs.setTreeAddress(idxTree);

        byte[] sigXmss = Xmss.sign(m, skSeed, idxLeaf, pkSeed, adrs, params, hash);
        byte[] sigHt = sigXmss;

        byte[] root = Xmss.pkFromSig(idxLeaf, sigXmss, m, pkSeed, adrs, params, hash);

        for (int j = 1; j < params.d; j++) {
            idxLeaf = (int)(idxTree & ((1L << params.hPrime) - 1));
            idxTree = idxTree >>> params.hPrime;

            adrs.setLayerAddress(j);
            adrs.setTreeAddress(idxTree);

            sigXmss = Xmss.sign(root, skSeed, idxLeaf, pkSeed, adrs, params, hash);
            sigHt = SlhUtils.concat(sigHt, sigXmss);

            if (j < params.d - 1) {
                root = Xmss.pkFromSig(idxLeaf, sigXmss, root, pkSeed, adrs, params, hash);
            }
        }
        return sigHt;
    }

    /**
     * ht_verify(M, sig_ht, PK.seed, idx_tree, idx_leaf, PK.root):
     * Verify a hypertree signature (Algorithm 16 in FIPS 205).
     */
    public static boolean verify(byte[] m, byte[] sigHt, byte[] pkSeed,
                                  long idxTree, int idxLeaf, byte[] pkRoot,
                                  SlhParams params, SlhHash hash) {
        int n = params.n;
        int xmssSigSize = (params.hPrime + params.wotsLen) * n;

        Address adrs = new Address();
        adrs.setLayerAddress(0);
        adrs.setTreeAddress(idxTree);

        byte[] sigXmss = SlhUtils.slice(sigHt, 0, xmssSigSize);
        byte[] node = Xmss.pkFromSig(idxLeaf, sigXmss, m, pkSeed, adrs, params, hash);

        for (int j = 1; j < params.d; j++) {
            idxLeaf = (int)(idxTree & ((1L << params.hPrime) - 1));
            idxTree = idxTree >>> params.hPrime;

            adrs.setLayerAddress(j);
            adrs.setTreeAddress(idxTree);

            sigXmss = SlhUtils.slice(sigHt, j * xmssSigSize, xmssSigSize);
            node = Xmss.pkFromSig(idxLeaf, sigXmss, node, pkSeed, adrs, params, hash);
        }

        return java.util.Arrays.equals(node, pkRoot);
    }
}
